### PR TITLE
refactor: make compatible with tiny-secp256k1@1.*

### PR DESCRIPTION
**Pull request by Hyunhum**
#55 shows the inconvenience to use bip32 because of the dependency on tiny-secp256k1@2.0.0+ including WASM, which requires some tricky config for app or web based engine.
As mentioned, if downgrade tiny-secp256k1 to version 1.1.6 not including WASM, it can be perfectly compatible with browser/app.
Though latest bip32 at that time was compatible with tiny-secp256k1@1.1.6, current version adds "testecc.js" which requires "xOnlyPointAddTweak" and "privateNegate" methods for tiny-secp256k1. As you know, tiny-secp256k1 v1 does not have those methods(like schnorr), so just wrap it up with if statement.
Tweak methods of BIP32 has a strong utility, but not necessarily used for most of dev who implements this library. It can be optional, to make lib more usable for devs.

### CODE REVIEW

**Comment by junderw** [comment]
1. You must also modify the ts files. JS and TS files must both be modified, so modify the TS file first, then run npm run build and commit the JS files too.
2. You should add a check inside tweakFromPublicKey that throws a nice error if ecc does not have tweakAdd and check inside tweakFromPrivateKey that throws a nice error if ecc does not have negate. "can't call undefined" error would be confusing.

**Comment by Hyunhum** [comment]
1 & 2 Done!

**Comment by junderw** [comment]
1. Fix trailing whitespace in 3 spots (lint error)
2. Fix format (run npm run format)
3. Adding branches lowered coverage. Add tests to make sure that running tweak() without the needed methods on ecc will throw an Error................ testing the testecc module is not needed tbh... but if you'd like to write a test that covers the if branching, feel free.

**Comment by Hyunhum** [comment]
Done. Test coverage is good now.

**Comment by landabaso** [comment]
Should the `TinySecp256k1Interface ` interface defined in this module match with the one defined in ecpair?

https://github.com/bitcoinjs/ecpair/blob/master/ts_src/ecpair.ts

Actually, they don't match. Not even before this PR.
`isPoint` is declared in ecpair and not in bip32.
And `pointAddScalar` is declared in bip32 and not in ecpair.
Maybe declare all of them and then make them optional?

I'm working on another module that will also have a Factory that receives an `TinySecp256k1Interface` and then passes it further to ecpair and bip32 factory functions and just realised about this.

I decided not to open an issue as it's kind of related to this PR. I was wondering whether it would make sense to unify them or it does not really matter.

**Comment by Hyunhum** [comment]
Both bip32 and ecpair receive parameter "ecc", which is usually from https://github.com/bitcoinjs/tiny-secp256k1.

I think it makes sense to unify interface and testecc because They use same ecc from tiny-secp256k1, but not sure whether necessary.

This PR suggests also some methods(which are not included in tiny-secp256k1 v1.1.6) of interface need to be optional, because v2 implements WASM which is experimental feature for web&app. Many client depends on v1.1.6.

**Comment by landabaso** [comment]
I totally understand your point and the reasoning behind this PR. Since it proposes changes to the `TinySecp256k1Interface`, it may be a good opportunity to consider if maintaining consistency in interfaces across different modules is desirable.

**Comment by junderw** [comment]
Still lost on what to do here. Any suggestions welcome.

**Comment by Hyunhum** [comment]
@landabaso @junderw  I checked again and "isPoint" is declared in [bip32](https://github.com/bitcoinjs/bip32/blob/master/types/bip32.d.ts#L50). 
Maybe "pointCompress" is what you intended? In ecpair, "pointCompress" is declared but not in bip32. I'm not really sure whether "pointCompress" is necessary for ecc of bip32(because it's not used in bip32, but used in ecpair). 
On the other hand, "pointAddScalar" used in bip32 so included, while not used in ecpair so excluded.
In my opinion, if it's not used, it doesn't have to be implemented. There are also other methods of tiny-secp256k1, which are excluded in bip32, "pointMultiply", "signRecoverable" and etc.
In conclusion, bip32 and ecpair does not need to unify tiny-secp256k1 interface, because each library uses ecc in a different way.
This PR is actually focused on browser/app compatibility(And it's done with current commits). Thus, maybe @landabaso you can make another new Issue or PR to deal with the problem you mentioned!

**Comment by landabaso** [comment]
Yes, `pointCompress` is not declared in ecpair and `pointAddScalar` is not declared in bip32.

> In conclusion, bip32 and ecpair does not need to unify tiny-secp256k1 interface, because each library uses ecc in a different way.

That's one possibility.

This can lead to TypeScript errors, as shown in the following example:

```javascript
import { TinySecp256k1Interface, ECPairFactory } from 'ecpair';

function myFactory(ecc: TinySecp256k1Interface) {
  const bip32 = BIP32Factory(ecc); //typescript error
  const ecpair = ECPairFactory(ecc);
}
```

I see these options:

1. Each module declares its own `TinySecp256k1Interface`, and the end user is responsible for dealing with these inconsistencies. This is the current setup, but it may not be ideal. I solved this by creating a new `TinySecp256k1Interface` with a superset of the other 2 interfaces.

2. All modules define the same `TinySecp256k1Interface` making sure to include all the functions documented in the [tiny-secp256k1 repository](https://github.com/bitcoinjs/tiny-secp256k1/blob/master/README.md). With no optional methods. Each module then tests only the subset of functions that are really required (and maybe also the other defined methods, even if not used). This approach is the easiest, solves the problem and is backwards compatible. However, it requires updating ecpair and bip32 (and any other relevant modules) every time tiny-secp256k1 adds a new method.

If you want to do 2), I can coordinate with @Hyunhum on a reference `TinySecp256k1Interface` and I can submit a pull request to ecpair. Option 1 (leave it as it is) is also fine. I don't want to be picky. This is just a minor issue.

**Comment by Hyunhum** [comment]
I'm just wondering, why with no optional?
And how about adding "pointCompress" to bip32 and "pointAddScalar" to ecpair only.
If type error concerns, this could be enough.
Adding all methods of tiny-secp256k1 will make both bip32 & ecpair heavily dependent on tiny-secp256k1 lib as you mention.

**Comment by Hyunhum** [comment]
@junderw Maybe no issue left to be this PR merged?
Each lib uses each one's TinySecp256k1Interface, or [tiny-secp256k1](https://github.com/bitcoinjs/tiny-secp256k1/pull/108) interface if that PR merged.

**Review by junderw — APPROVED** [review]
